src/ir/possible-contents.cpp

```diff
@@ -0,0 +1,1610 @@
+        Index j = 0;
+        for (auto t : curr->getLocalType(i)) {
+          if (t.isDefaultable()) {
+            info.links.push_back(
+              {getNullLocation(t), LocalLocation{curr, i, j}});
+          }
+          j++;
+        }
+      }
+    }
+
+    // Functions with a result can flow a value out from their body.
+    addResult(curr->body);
+
+    // See visitPop().
+    assert(handledPops == totalPops);
+  }
+
+  // Helpers
+
+  // Handles the value sent in a break instruction. Does not handle anything
+  // else like the condition etc.
+  void handleBreakValue(Expression* curr) {
+    BranchUtils::operateOnScopeNameUsesAndSentValues(
+      curr, [&](Name target, Expression* value) {
+        if (value) {
+          for (Index i = 0; i < value->type.size(); i++) {
+            // Breaks send the contents of the break value to the branch target
+            // that the break goes to.
+            info.links.push_back({ExpressionLocation{value, i},
+                                  BranchLocation{getFunction(), target, i}});
+          }
+        }
+      });
+  }
+
+  // Handles receiving values from breaks at the target (as in a block).
+  void handleBreakTarget(Expression* curr) {
+    if (isRelevant(curr->type)) {
+      BranchUtils::operateOnScopeNameDefs(curr, [&](Name target) {
+        for (Index i = 0; i < curr->type.size(); i++) {
+          info.links.push_back({BranchLocation{getFunction(), target, i},
+                                ExpressionLocation{curr, i}});
+        }
+      });
+    }
+  }
+
+  // Connect a child's value to the parent, that is, all content in the child is
+  // now considered possible in the parent as well.
+  void receiveChildValue(Expression* child, Expression* parent) {
+    if (isRelevant(parent) && isRelevant(child)) {
+      // The tuple sizes must match (or, if not a tuple, the size should be 1 in
+      // both cases).
+      assert(child->type.size() == parent->type.size());
+      for (Index i = 0; i < child->type.size(); i++) {
+        info.links.push_back(
+          {ExpressionLocation{child, i}, ExpressionLocation{parent, i}});
+      }
+    }
+  }
+
+  // See the comment on CollectedInfo::childParents.
+  void addSpecialChildParentLink(Expression* child, Expression* parent) {
+    if (isRelevant(child->type)) {
+      info.childParents[child] = parent;
+    }
+  }
+
+  // Adds a root, if the expression is relevant. If the value is not specified,
+  // mark the root as containing Many (which is the common case, so avoid
+  // verbose code).
+  void addRoot(Expression* curr,
+               PossibleContents contents = PossibleContents::many()) {
+    if (isRelevant(curr)) {
+      addRoot(ExpressionLocation{curr, 0}, contents);
+    }
+  }
+
+  // As above, but given an arbitrary location and not just an expression.
+  void addRoot(Location loc,
+               PossibleContents contents = PossibleContents::many()) {
+    info.roots.emplace_back(loc, contents);
```

Comment:
Can this`roots` be duplicated?

Reply:
Do you mean can it contain duplicate values? There wouldn't be any harm if there were, but I think there can't be, since we walk the code and add roots as we find things, and we don't walk things twice.